### ISSUE TITLE
Add support for incremental annotation processing with Gradle 4.7+

### ISSUE THREAD

**Issue by lucamtudor** [objectbox/objectbox-java#620]
Gradle 4.7 added support for incremental annotation processing, docs [here](https://docs.gradle.org/nightly/userguide/java_plugin.html#sec:incremental_annotation_processing).

For additional help, follow up with the [gradle dedicated issue](https://github.com/gradle/gradle/issues/5277).

**Comment by greenrobot-team** [comment]
Thanks. Already looked into this a while ago and the issue is that our processor does not really fit either mode. `@Entity` classes cause generation of multiple files (Entity Info, Cursors), but all entity classes are also aggregated in a single class (`MyObjectBox`).

In addition for relations multiple `@Entity` classes may influence the contents of a generated class. This does not seem to fit "isolating" mode.

The ObjectBox processor also can't use "aggregating" mode because it needs access to parameter names (would require a compiler flag with that mode).

Side note: when using the Android Plugin it is still unclear (last time I checked) how to do incremental annotation processing anyhow.
-ut

**Comment by lucamtudor** [comment]
I'm afraid I can't help, I'm no gradle wizard.

**Comment by greenrobot-team** [comment]
News: with Android Gradle Plugin 3.3.0 incremental annotation processor support is available for Java-only projects. But not for Kotlin or mixed Kotlin+Java projects.

Also, there is a note that it does not work with Gradle 5.1 and higher.
https://developer.android.com/studio/releases/gradle-plugin

So: nothing to do here for now.
-ut

**Comment by dbriggsDFleet** [comment]
Can this be unblocked with the release of Android Gradle Plugin 3.5.0?  "Additionally, KAPT 1.3.30 and higher also support incremental annotation processors, which you can enable by including kapt.incremental.apt=true in your gradle.properties file."
https://developer.android.com/studio/releases/gradle-plugin#3-5-0

**Comment by greenrobot-team** [comment]
@dbriggsDFleet Sure, but the biggest, unresolved issues are still those described in https://github.com/objectbox/objectbox-java/issues/620#issuecomment-444025267.

**Comment by jrcacd** [comment]
This is a must.
With incremental annotation processing we could do really fast unit testing with objectBox, as we would have to wait a lot less for the project to compile

**Comment by greenrobot-team** [comment]
We just released `2.5.1` which supports incremental annotation processing. It is turned **off by default**.

To turn it on set `objectbox.incremental` to `true` in `build.gradle`:
```
android {
    defaultConfig {
        javaCompileOptions {
            annotationProcessorOptions {
                arguments = [ "objectbox.incremental" : "true" ]
            }
        }
    }
}
```
We appreciate your feedback on this feature.

If all goes well, incremental annotation processing will be on by default in a future release.

**Comment by sebhard** [comment]
Great, will try it out asap!

**Comment by greenrobot** [comment]
If you see improved build times, we'd be very happy if you could share some numbers from your project (e.g. build times before and after).

**Comment by greenrobot-team** [comment]
`3.0.0-alpha1` has incremental annotation processing enabled by default.
https://docs.objectbox.io/#v-3-0-0-alpha1-2020-03-09

**Comment by greenrobot-team** [comment]
2.6.0-RC has incremental annotation processing enabled by default as well.
https://docs.objectbox.io/#v-2-6-0-rc-2020-04-28

**Comment by greenrobot-team** [comment]
With `2.6.0` incremental annotation processing is now enabled by default.
https://docs.objectbox.io/#v-2-6-0-2020-06-09

**Comment by emartynov** [comment]
Hi people,

can we make other task incremental?

<img width="1064" alt="Screenshot 2020-07-08 at 15 44 48" src="https://user-images.githubusercontent.com/295424/86926230-1b5fa500-c132-11ea-9a51-e2e91192d05c.png">

Sure it is matter of 0 seconds to run. But this always distracting attention when you checking issues with the build.

**Comment by greenrobot** [comment]
@emartynov can you please open a new issue for this? Thanks

**Comment by greenrobot-team** [comment]
@emartynov Thanks! I guess you did not get around creating an issue for this (couldn't find one)? Anyhow, this is already fixed internally and will ship with the next update.

**Comment by emartynov** [comment]
Hi people, you're early starters. You're right I didn't. Thank you for the update!
